source/extensions/filters/listener/original_src/original_src.cc

```diff
@@ -0,0 +1,44 @@
+#include "extensions/filters/listener/original_src/original_src.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/network/connection.h"
+
+#include "common/common/assert.h"
+#include "common/network/utility.h"
+
+#include "extensions/filters/listener/original_src/original_src_socket_option.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace OriginalSrc {
+
+OriginalSrcFilter::OriginalSrcFilter(const Config& config) : config_(config) {}
+
+Network::FilterStatus OriginalSrcFilter::onAccept(Network::ListenerFilterCallbacks& cb) {
+  auto& socket = cb.socket();
+  auto address = socket.remoteAddress();
+  ASSERT(address);
+
+  ENVOY_LOG(trace, "Got a new connection in the original_src filter for address {}",
+            address->asString());
+
+  if (address->type() != Network::Address::Type::Ip) {
+    // nothing we can do with this.
+    return Network::FilterStatus::Continue;
+  }
+
+  auto address_without_port = Network::Utility::getAddressWithPort(*address, 0);
+
+  Network::Socket::OptionConstSharedPtr new_option =
+      std::make_shared<OriginalSrcSocketOption>(std::move(address_without_port));
```

Comment:
Should pass `config_.mark_` to the option here?

Reply:
I will when I do the SO_MARK option work (again, a future PR, probably my next one, since it's necessary for this to be useful). :)

I have the work for SO_MARK languishing in another PR that I'll probably cancel. I'll cherry-pick the stuff I care about into a new one.